src/databricks/labs/ucx/azure/credentials.py

```diff
@@ -0,0 +1,303 @@
+                read_only=read_only,
+            )
+            return StorageCredentialValidationResult.from_validation(
+                storage_credential, validation, sp.permission_mapping.prefix
+            )
+        except InvalidParameterValue:
+            logger.warning(
+                "There is an existing external location overlaps with the prefix that is mapped to the service principal and used for validating the migrated storage credential. Skip the validation"
+            )
+            return StorageCredentialValidationResult.from_validation(
+                storage_credential,
+                ValidateStorageCredentialResponse(
+                    is_dir=None,
+                    results=[
+                        ValidationResult(
+                            message="The validation is skipped because an existing external location overlaps with the location used for validation."
+                        )
+                    ],
+                ),
+                sp.permission_mapping.prefix,
+            )
+
+
+class ServicePrincipalMigration:
+
+    def __init__(
+        self,
+        installation: Installation,
+        ws: WorkspaceClient,
+        resource_permissions: AzureResourcePermissions,
+        sp_crawler: AzureServicePrincipalCrawler,
+        storage_credential_manager: StorageCredentialManager,
+    ):
+        self._output_file = "azure_service_principal_migration_result.csv"
+        self._installation = installation
+        self._ws = ws
+        self._resource_permissions = resource_permissions
+        self._sp_crawler = sp_crawler
+        self._storage_credential_manager = storage_credential_manager
+
+    @classmethod
+    def for_cli(cls, ws: WorkspaceClient, prompts: Prompts, product='ucx'):
+        if not ws.config.is_azure:
+            logger.error("Workspace is not on azure, please run this command on azure databricks workspaces.")
+            raise SystemExit()
+
+        msg = (
+            "Have you reviewed the azure_storage_account_info.csv "
+            "and confirm listed service principals are allowed to be checked for migration?"
+        )
+        if not prompts.confirm(msg):
+            raise SystemExit()
+
+        installation = Installation.current(ws, product)
+        config = installation.load(WorkspaceConfig)
+        sql_backend = StatementExecutionBackend(ws, config.warehouse_id)
+        azurerm = AzureResources(ws)
+        locations = ExternalLocations(ws, sql_backend, config.inventory_database)
+
+        resource_permissions = AzureResourcePermissions(installation, ws, azurerm, locations)
+        sp_crawler = AzureServicePrincipalCrawler(ws, sql_backend, config.inventory_database)
+
+        storage_credential_manager = StorageCredentialManager(ws)
+
+        return cls(installation, ws, resource_permissions, sp_crawler, storage_credential_manager)
+
+    def _read_databricks_secret(self, scope: str, key: str, application_id: str) -> str | None:
```

Comment:
I've created a SecretMixin class in #950, so you reuse/update this method from there

Reply:
great, I applied SecretMixin. I also let `_get_secret_if_exists` to catch the UnicodeDecodeError just in case, please have a look.